src/lexer/strings.js

```diff
@@ -126,25 +129,100 @@ module.exports = {
   },
 
   // check if its a DOC end sequence
-  isDOC_MATCH: function() {
+  isDOC_MATCH: function(offset, consumeLeadingSpaces) {
     // @fixme : check if out of text limits
+
+    // consumeLeadingSpaces is false happen DOC prematch END HEREDOC stage.
+
+    // Ensure current state is really after a new line break, not after a such as ${variables}
+    const prev_ch = this._input.substring(offset - 2, offset - 1);
+    if (prev_ch !== "\n" && prev_ch !== "\r") {
+      return false;
+    }
+
+    // skip leading spaces or tabs
+    let indentation_uses_spaces = false;
+    let indentation_uses_tabs = false;
+    // reset heredoc_label structure
+    let indentation = 0;
+    let leading_ch = this._input.substring(offset - 1, offset);
+
+    while (/[\t ]/.test(leading_ch)) {
+      if (leading_ch === " ") {
+        indentation_uses_spaces = true;
+      } else if (leading_ch === "\t") {
+        indentation_uses_tabs = true;
+      }
+
+      leading_ch = this._input[offset + indentation];
+      indentation++;
+    }
```

Comment:
I think we should throw error on invalid indentation like:
```
// Parse error: Invalid body indentation level (expecting an indentation at least 5) in %s on line %d
```

How in rfc

Reply:
Get it. I will added it later

Reply:
> I think we should throw error on invalid indentation like:
> 
> ```
> // Parse error: Invalid body indentation level (expecting an indentation at least 5) in %s on line %d
> ```
> 
> How in rfc

Just added code to detect this kind of error and raise the error via `this.raiseError` , I am not sure if I handle it in the right way. please help check it.

Reply:
It's ok, but you should also test the code with `suppressErrors` mode in order to check it's fallback mode